igvc_ws/src/igvc_serial/src/serial_node.py

```diff
@@ -38,39 +42,83 @@ def run(self):
+        # semi-regularly checked. This is better than rospy.Rate because it will continously wait for new message
+        # instead of only checking on a fixed interval.
+        self.serial_obj.timeout = 1
+
+        # Assumes String type for now. This class will need to be adapted in the future for different message types.
+        self.publisher = rospy.Publisher(topic, gps, queue_size=1)
+
+    def run(self):
+        while not rospy.is_shutdown():
+            coord = self.serial_obj.readline() # Assume all messages end in newline character. This is standard among SCR IGVC serial messages.
+
+            try:
+                if coord:
+                    coord_json = json.loads(coord)
+
+                    coord_msg = gps()
+                    coord_msg.latitude = coord_json['latitude']
+                    coord_msg.longitude = coord_json['longitude']
+                    coord_msg.hasSignal = coord_json['hasSignal']
+
+                    self.publisher.publish(coord_msg)
+            except ValueError:
+                pass
 
+# Constructs motor message from given data and sends to serial
 def motors_out(data):
+
+    # JSON packet to be sent
     motion_pkt = {
         "motorLeft": data.left,
         "motorRight": data.right,
     }
 
-    json_dump = json.dumps(motion_pkt, separators=(',', ':'))
-    # print("writing " + json_dump)
-    out = (json_dump + "\n").encode() # encode json string to bytes
+    # Encode JSON string to bytes
+    json_dump = json.dumps(motion_pkt, separators = (',', ':'))
+    out = (json_dump + "\n").encode()
 
     serials["motor"].write(out)
 
-def serial_node():
-
-    # Setup node
-    rospy.init_node("serial_node", anonymous = True)
 
-    # Create Serial objects to read and write from serial devices
-    serials["motor"] = serial.Serial(port = '/dev/igvc-nucleo-120', baudrate = 115200)
+# Initialize the serial node
+# Node handles all serial communication within the robot (motor, GPS)
+def init_serial_node():
+    
+    # Setup serial node
+    rospy.init_node("serial_node", anonymous = False)
 
-    # Subscribe to necessary topics
-    rospy.Subscriber("/igvc/motors_raw", motors, motors_out)
+    # Setup motor serial and subscriber
+    motor_serial = serials["motor"] = serial.Serial(port = '/dev/igvc-nucleo-120', baudrate = 115200)
+    rospy.Subscriber('/igvc/motors_raw', motors, motors_out)
     
-    # # The following lines are an example of how to create a serial reading object thread
-    motor_response_thread = SerialReadThread(serial_obj = serials["motor"], topic = '/igvc/velocity')
+    motor_response_thread = VelocitySerialReadThread(serial_obj = serials["motor"], topic = '/igvc/velocity')
     motor_response_thread.start()
 
+    # Setup GPS serial and publisher
+    gps_serial = serials["gps"] = serial.Serial(port = '/dev/ttyACM0', baudrate = 9600)
+
+    gps_response_thread = GPSSerialReadThread(serial_obj = serials["gps"], topic = '/igvc/gps')
```

Comment:
This is just a general comment, but the design pattern of this file confuses me. Are we trying to have a thread for each sensor, or a thread for each serial I/O we have? I don't think having a thread for each sensor is needed if there are sensors that share a serial bus (obviously not a problem with the GPS, as it is the sole sensor on its board right now). Also, if we are doing a thread for each serial channel we need to add a thread for outputting values to the motors.

This isn't a problem with this code change, but I'd like to know what's happening here @noahzemlin